text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,606 @@
+
+A stack can provide stackpacks by including them in the `/cnb/stack/buildpacks` directory, and providing an `/cnb/stack/order.toml` (following the [`order.toml` schema](https://github.com/buildpacks/spec/blob/main/platform.md#ordertoml-toml)) to define their order of execution. The order can be overridden in the `builder.toml` with the following configuration:
+
+```
+[[stack.order]]
+  [[stack.order.group]]
+    id = "<stackpack id>"
+```
+
+A stackpack will only execute if it passes detection. When the stackpack is executed, its detect and build scripts use the same parameters as the regular buildpacks. But the arguments may differ. For example, the first positional argument to the `bin/build` (the `<layers>` directory) MUST NOT be a standard layers directory adhering to the [Buildpacks Build specification](https://github.com/buildpacks/spec/blob/main/buildpack.md#build).
+
+The stackpack's snapshot layer may be enriched by writing a `launch.toml` file. The `launch.toml` may define globs of files to be excluded from the image when it is _used_. Any excluded path may also be marked as _cached_, so that those excluded paths are recovered when the image is _used_. The term _used_ is defined as:
+
+* *Used for build-time build*: A given path is excluded at normal buildpack build-time, and recovered the next time the build image is extended with the stackpack.
+* *Used for build-time run*: A given path is excluded from the final image, and recovered the next time the run image is extended with the stackpack (either rebase or rebuild).
+* *Used for rebase run*: A given path is excluded from the rebased image, and recovered the next time the run image is extended with the stackpack (either rebase or rebuild).
+
+For example, a stack buildpack may choose to exclude `/var/cache` from the final run image, but may want to mark it as _cached_ to have it restored during build-time and rebase.
+
+## Mixins
+
+### Providing Mixins
+
+A stack buildpack MAY define a set of mixins it provides in either of two ways:
+
+1. Statically in the `buildpack.toml`
+1. Dynamically in the Build Plan.
+
+A stack buildpack MAY define a set of stack specific mixins in the `buildpack.toml` with the following schema:
+
+```
+[[stacks]]
+id = "<stack id or *>"
+
+[stacks.provides]
+any = <boolean (default=false)>
+mixins = [ "mixin name" ]
+```
+
+Additionally, mixins MAY be dynamically provided in the build plan:
+
+```
+[[provides]]
+name = "<mixin name>"
+any = <boolean (default=false)>
+mixin = <boolean (default=false)>
+```
+
+A userspace buildpack MAY require mixins in the build plan
+
+```
+[[requires]]
+name = "<mixin name>"
+mixin = <boolean (default=false)>
+```
+
+A userspace buildpack MAY NOT provide mixins in the build plan.
+
+### Requiring Mixins
+
+Stack buildpacks MUST NOT require any entries in the build plan (neither mixins nor non-mixins). This ensures that we do not need to re-run the detect phase.
```

Comment:
I don't remember, but I think this had to do with something @jabrown85 ran into while prototyping the mixin resolution

Reply:
I don't think we'd need to run detect twice, but we can't allow stackpacks to require `yj` from normal buildpacks. Otherwise you can't run `extend` and `build` in parallel and it means you have to run the stackpack after the buildpack that provides `yj`